### PR TITLE
Added generation counter in content tree config

### PR DESCRIPTION
**Pull request by zed-rishabh**
For container images, we won't get SHA in config so we need to resolve it before doing download/verification in EVE. With the help of the generation counter, we will identify whether we want to resolve SHA or not.
generation counter needs to be an array because multiple instances can run on a different version of an image corresponding to the same content-id (in case of containers).
cc: @bharani-zededa @nordmark

### CODE REVIEW

**Comment by zed-rishabh** [comment]
> Couldn't the controller create a new ContentTree UUID and send it to EVE?

If we will get a new ContentTree UUID then we don't need an array.

> Also, don't know if we should report the generation counter in the ZVolumeInfo message
> If we need one for CT then also report for the CT info

Ok, I will add it in info as well.

**Comment by zed-rishabh** [comment]
@bharani-zededa If two instances want to run on same image say ```nginx:latest```, will we get one ContentTree UUID or two?

**Comment by bharani-zededa** [comment]
> @bharani-zededa If two instances want to run on same image say `nginx:latest`, will we get one ContentTree UUID or two?

That depends on the controller implementation (and subject to change) depending on user work flow. EVE should not make an assumption either way.

FWIW, I don't think this should be an array (and if it is an array, I don't understand how controller can provide that information).

File: api/proto/config/storage.proto [hunk 1]
```diff
@@ -126,6 +126,14 @@ message ContentTree {
   uint64 maxSizeBytes = 6; // used for capping resource consumption in EVE
   SignatureInfo siginfo = 7;
   string displayName = 8; // Optional friendly name echo'ed in info
+
+  // For container ContentTree, sha256 won't be available in config
```
Comment (eriknordmark) [inline api/proto/config/storage.proto:130]: Might not be available; we can still have a CT specified with a sha256, right?

File: api/proto/config/storage.proto [hunk 2]
```diff
@@ -126,6 +126,14 @@ message ContentTree {
...
+  // So, sha256 needs to be resolved corresponding to the tag
```
Comment (eriknordmark) [inline api/proto/config/storage.proto:131]: If the sha256 is not, then ...

File: api/proto/info/info.proto [hunk 3]
```diff
@@ -539,8 +539,9 @@ message ZInfoVolume {
   VolumeResources resources = 4;
   ZSwState state = 5;
   uint32 progressPercentage = 6; // 0-100% Use depends on state i.e., download vs. verification vs. volume creation
+  int64 generation_count = 7; // version of volume
 
-  ErrorInfo volumeErr = 7;
+  ErrorInfo volumeErr = 8;
```
Comment (eriknordmark) [inline api/proto/info/info.proto:544]: Is this already used in the controller? If so, it is better to not change the tag but to make the generation_count be 8

File: api/proto/info/info.proto [hunk 4]
```diff
@@ -559,9 +560,10 @@ message ZInfoContentTree {
   UsageInfo usage = 5;
   ZSwState state = 6;   // INITIAL, DOWNLOAD_STARTED, DOWNLOADED, DELIVERED
   uint32 progressPercentage = 7; // 0-100% Depending on state if it is download or verification
+  int64 generation_count = 8; // version of content tree
 
-  ErrorInfo err = 8;
-  repeated string componentShaList = 9; // sha's for the referenced blobs
+  ErrorInfo err = 9;
+  repeated string componentShaList = 10; // sha's for the referenced blobs
```
Comment (eriknordmark) [inline api/proto/info/info.proto:566]: Ditto

File: api/proto/config/storage.proto (hunk 1 above)
Comment (zed-rishabh) [reply api/proto/config/storage.proto]: Yes. Resolved.

File: api/proto/config/storage.proto (hunk 2 above)
Comment (zed-rishabh) [reply api/proto/config/storage.proto]: Resolved

File: api/proto/info/info.proto (hunk 3 above)
Comment (zed-rishabh) [reply api/proto/info/info.proto]: Resolved

File: api/proto/info/info.proto (hunk 4 above)
Comment (zed-rishabh) [reply api/proto/info/info.proto]: Resolved

File: api/proto/config/storage.proto [hunk 5]
```diff
@@ -131,6 +131,14 @@ message ContentTree {
   uint64 maxSizeBytes = 6;
   SignatureInfo siginfo = 7;
   string displayName = 8; // Optional friendly name echo'ed in info
+
+  // For container ContentTree, sha256 might not be available in config.
```
Comment (rvs) [inline api/proto/config/storage.proto:135]: While I'm not against this improvement -- I must point out that I'm moving closer to rethinking what SHAs do we use where (that's part of the discussion we're having with @deitch and some upstream containerd folks). Expect a proposal sometime soon that may streamline this area.

Comment (rvs) [reply api/proto/config/storage.proto:135]: I guess this is mostly a comment for @eriknordmark and @adarsh-zededa

**Pull request merged**